### ISSUE TITLE
Add metadata to functions script side

### ISSUE THREAD

**Issue by stargazing-dino** [rhaiscript/rhai#925]
Hi, is there a way to add metadata to Rhai functions? Kind of similar to Python's decorator syntax.

```py
@app.route("/users/<id>")
def get_user(id):
    return fetch_user(id)
```

Or even Rust's macros (example from Actix)

```rs
#[get("/")]
async fn hello() -> impl Responder {
    HttpResponse::Ok().body("Hello world!")
}
```

The examples I gave might be doing more than what I'm after or working in a completely different way. For my use case I don't need anything other than marking certain functions with with arbitrary attributes (perhaps with data).

One workaround I was thinking of was maybe forcing the user to make an object, defining an object with the function as a closure and my metadata on a field but it looks kinda odd:

```rhai
let my_function = #{
    metadata: 42,
    function: || // do_something_cool
};
```

and I'd wish there was a way to just do something like:

```rhai
@(my_data(42))
fn call_me() {
    return 3;
}
```

Is this possible through custom syntax perhaps or even now in some kind of work around way?

**Comment by schungx** [comment]
You can always access metadata in Rust.  So I suppose you want to access metadata in code?

https://rhai.rs/book/engine/metadata/index.html

**Comment by stargazing-dino** [comment]
Yes, I meant specifically access metadata of a function in Rhai code.

On a complexity scale how hard would implement something like this directly into Rhai be?

In the meantime, I'll see if I can't get by with the doc comments alternative - thank you for the suggestion:)

**Comment by schungx** [comment]
Quite complex as it would require extending the syntax. Then we'll need to design the syntax to make sure it is flexible enough.

**Comment by schungx** [comment]
If you use doc comments, it is built in.

But you have to parse the MarkDown. You can use a crate for it.

**Comment by stargazing-dino** [comment]
The markdown idea is definitely interesting! It might be more than what I want for my usecase though. I haven't tried it yet but I think I'll likely do something like this:

```rhai
fn call_me() {
    return 3;
}

let extra_data = Route("/");

register(call_me, extra_data);
```

And the register function could do all the fancy logic. Its signature could take the Fn_Ptr and a Dynamic. A little more setup but I think it should cover my immediate needs. I wonder, would I be able to do something like this?

```rhai
call_me.register(extra_data);
```

(anyways don't mind me too much, I'm going to play with all this this week)

**Comment by schungx** [comment]
> call_me.register(extra_data)

Yes you can. Just register a method for the `FnPtr` type.

**Comment by schungx** [comment]
Closing this for now.  Feel free to reopen.
